diff --git a/test/test_main.cc b/test/test_main.cc
index 82b39e48..fb2db610 100644
--- a/test/test_main.cc
+++ b/test/test_main.cc
@@ -11,7 +11,9 @@ extern "C" {
     struct cpu_features test_cpu_features;
 }

-GTEST_API_ int main(int argc, char **argv) {
+GTEST_API_ int main(int argc, char **argv);
+
+int main(int argc, char **argv) {
   printf("Running main() from %s\n", __FILE__);
   cpu_check_features(&test_cpu_features);
   testing::InitGoogleTest(&argc, argv);
